reconcile/templating/renderer.py

```diff
@@ -0,0 +1,179 @@
+                "w",
+                encoding="utf-8",
+            ) as f:
+                f.write(output.content)
+
+    def read(self, path: str) -> Optional[str]:
+        try:
+            with open(
+                f"{join_path(self.app_interface_data_path, path)}",
+                "r",
+                encoding="utf-8",
+            ) as f:
+                return f.read()
+        except FileNotFoundError:
+            logging.debug(f"File not found: {path}, need to create it")
+
+
+def unpack_static_variables(
+    collection_variables: TemplateCollectionVariablesV1,
+) -> dict:
+    return collection_variables.static or {}
+
+
+def unpack_dynamic_variables(
+    collection_variables: TemplateCollectionVariablesV1, gql: gql.GqlApi
+) -> dict:
+    if not collection_variables.dynamic:
+        return {}
+    return {
+        dv.name: list(gql.query(dv.query).values())
+        for dv in collection_variables.dynamic or []
+    }
+
+
+class TemplateRendererIntegrationParams(PydanticRunParams):
+    app_interface_data_path: Optional[str]
+
+
+def join_path(base: str, sub: str) -> str:
+    # not using os.path.sep, since app-interface relies on unix paths
+    if sub.startswith(APP_INTERFACE_PATH_SEPERATOR):
+        return os.path.join(base, sub[1:])
+    return os.path.join(base, sub)
+
+
+class TemplateRendererIntegration(QontractReconcileIntegration):
+    def __init__(self, params: TemplateRendererIntegrationParams) -> None:
+        super().__init__(params)
+
+    @staticmethod
+    def process_template(
+        template: TemplateV1,
+        variables: dict,
+        persistence: FilePersistence,
+        ruaml_instance: yaml.YAML,
+    ) -> Optional[TemplateOutput]:
+        r = create_renderer(
+            template,
+            TemplateData(
+                variables=variables,
+            ),
+        )
+        target_path = r.render_target_path()
+
+        current_str = persistence.read(
+            target_path,
+        )
+        if current_str is None and template.patch:
+            raise ValueError(f"Can not patch non-existing file {target_path}")
+
+        if current_str:
+            r.data.current = ruaml_instance.load(current_str)
+
+        if r.render_condition():
+            output = r.render_output()
+
+            if current_str != output:
+                print(f"diff in template {template.name} for target_path {target_path}")
+                return TemplateOutput(
+                    path=target_path,
+                    content=output,
+                )
+        return None
+
+    @property
+    def name(self) -> str:
+        return QONTRACT_INTEGRATION
+
+    def run(self, dry_run: bool) -> None:
+        outputs: list[TemplateOutput] = []
+        gql_no_validation = init_from_config(validate_schemas=False)
+        persistence = LocalFilePersistence(self.params.app_interface_data_path)
+
+        ruaml_instance = yaml.YAML()
```

Comment:
Please do me a favor and move this part into a utils method. There are already 3 different places where we initialize `rumal` like this.

Reply:
Yes, that's a good idea. I'll do it in a new PR